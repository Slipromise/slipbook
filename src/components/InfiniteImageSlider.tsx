import React, { useCallback, useMemo, useState } from "react";
import styles from "@/styles/components/InfiniteImageSlider.module.scss";
import Image, { StaticImageData } from "next/image";
import { IoIosArrowDropleft, IoIosArrowDropright } from "react-icons/io";
import { RxLoop } from "react-icons/rx";
import { useDebounce, useInterval } from "react-use";
import lodash, { debounce } from "lodash";
type Props = {
  images: (StaticImageData | string)[];
  isAuto?: boolean;
  autoFrequency?: number;
  onAuto?: () => void;
};

function InfiniteImageSlider({
  images,
  onAuto,
  isAuto,
  autoFrequency = 3000,
}: Props) {
  const [offset, setOffset] = useState(() => 0);

  const [inAction, setInAction] = useState(() => false);

  const indexes = useMemo<number[]>(() => {
    var result = [];
    for (let i = 0; i < 7; i++) {
      result[i] = offset + i;
    }
    return result;
  }, [offset]);

  const onNext = useCallback(() => {
    setOffset((prev) => (prev - 1 <= Number.MIN_SAFE_INTEGER ? 0 : prev - 1));
  }, []);

  const onPrev = useCallback(() => {
    setOffset((prev) => (prev + 1 >= Number.MAX_SAFE_INTEGER ? 0 : prev + 1));
  }, []);

  useDebounce(
    () => {
      setInAction(false);
    },
    300,
    [offset]
  );

  useInterval(
    onNext,
    isAuto && !inAction && autoFrequency > 200 ? autoFrequency : null
  );

  if (images.length <= 1) {
    return null;
  }

  return (
    <div className={styles.container}>
      {indexes.map((index) => (
        <div className={styles.item} key={index}>
          <Image
            src={
              index >= 0 || Math.abs(index % images.length) === 0
                ? images[index % images.length]
                : images[Math.abs(images.length + (index % images.length))]
            }
            alt=""
          />
        </div>
      ))}
      <div className={styles.buttons}>
        <IoIosArrowDropleft
          onClick={() => {
            onPrev();
            setInAction(true);
          }}
        ></IoIosArrowDropleft>
        <IoIosArrowDropright
          onClick={() => {
            onNext();
            setInAction(true);
          }}
        ></IoIosArrowDropright>
        <RxLoop onClick={onAuto} data-on={isAuto} />
      </div>
    </div>
  );
}

export default InfiniteImageSlider;
