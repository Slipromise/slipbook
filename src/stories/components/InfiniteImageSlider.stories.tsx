import InfiniteImageSlider from "@/components/InfiniteImageSlider";
import { useArgs } from "@storybook/preview-api";
import type { Meta, StoryObj } from "@storybook/react";
import image1 from "../assets/00001-2086282684.png";
import image2 from "../assets/00002-442336389.png";
import image3 from "../assets/00003-3958138843.png";
import image4 from "../assets/00004-15481453.png";
import image5 from "../assets/00005-15481453.png";
import image6 from "../assets/00006-15481453.png";
import image7 from "../assets/00007-15481453.png";

const meta: Meta<typeof InfiniteImageSlider> = {
  component: InfiniteImageSlider,
  tags: ["autodocs"],
  parameters: {
    layout: "centered",
  },
};

const images = [image1, image2, image3, image4, image5, image6, image7];

export default meta;

type Story = StoryObj<typeof InfiniteImageSlider>;

export const Normal: Story = {
  args: {
    images,
  },
  render: function Render(args) {
    const [{ isAuto }, updateArgs] = useArgs<{ isAuto?: boolean }>();

    function onChange() {
      updateArgs({ isAuto: !isAuto });
    }

    return <InfiniteImageSlider {...args} onAuto={onChange} isAuto={isAuto} />;
  },
};

export const AutoLoop: Story = {
  ...Normal,
  args: {
    images,
    isAuto: true,
  },
};

export const LessImage: Story = {
  ...Normal,
  args: {
    images: [images[0], images[1]],
    isAuto: true,
  },
};
