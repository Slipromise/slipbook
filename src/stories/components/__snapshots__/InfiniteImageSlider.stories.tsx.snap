// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stories/components/InfiniteImageSlider AutoLoop smoke-test 1`] = `
<div class="_container_1r3be_1">
  <div class="_item_1r3be_11">
    <img src="/src/stories/assets/00001-2086282684.png"
         alt
    >
  </div>
  <div class="_item_1r3be_11">
    <img src="/src/stories/assets/00002-442336389.png"
         alt
    >
  </div>
  <div class="_item_1r3be_11">
    <img src="/src/stories/assets/00003-3958138843.png"
         alt
    >
  </div>
  <div class="_item_1r3be_11">
    <img src="/src/stories/assets/00004-15481453.png"
         alt
    >
  </div>
  <div class="_item_1r3be_11">
    <img src="/src/stories/assets/00005-15481453.png"
         alt
    >
  </div>
  <div class="_item_1r3be_11">
    <img src="/src/stories/assets/00006-15481453.png"
         alt
    >
  </div>
  <div class="_item_1r3be_11">
    <img src="/src/stories/assets/00007-15481453.png"
         alt
    >
  </div>
  <div class="_buttons_1r3be_59">
    <svg stroke="currentColor"
         fill="currentColor"
         stroke-width="0"
         viewbox="0 0 512 512"
         height="1em"
         width="1em"
         xmlns="http://www.w3.org/2000/svg"
    >
      <path d="M301.3 147.6c-7.5-7.5-19.8-7.5-27.3 0l-95.4 95.7c-7.3 7.3-7.5 19.1-.6 26.6l94 94.3c3.8 3.8 8.7 5.7 13.7 5.7 4.9 0 9.9-1.9 13.6-5.6 7.5-7.5 7.6-19.7 0-27.3l-79.8-81 81.9-81.1c7.5-7.5 7.5-19.7-.1-27.3z">
      </path>
      <path d="M256 48C141.1 48 48 141.1 48 256s93.1 208 208 208 208-93.1 208-208S370.9 48 256 48zm124.4 332.4C347.2 413.7 303 432 256 432s-91.2-18.3-124.4-51.6C98.3 347.2 80 303 80 256s18.3-91.2 51.6-124.4C164.8 98.3 209 80 256 80s91.2 18.3 124.4 51.6C413.7 164.8 432 209 432 256s-18.3 91.2-51.6 124.4z">
      </path>
    </svg>
    <svg stroke="currentColor"
         fill="currentColor"
         stroke-width="0"
         viewbox="0 0 512 512"
         height="1em"
         width="1em"
         xmlns="http://www.w3.org/2000/svg"
    >
      <path d="M210.7 147.6c7.5-7.5 19.8-7.5 27.3 0l95.4 95.7c7.3 7.3 7.5 19.1.6 26.6l-94 94.3c-3.8 3.8-8.7 5.7-13.7 5.7-4.9 0-9.9-1.9-13.6-5.6-7.5-7.5-7.6-19.7 0-27.3l79.9-81.1-81.9-81.1c-7.6-7.4-7.6-19.6 0-27.2z">
      </path>
      <path d="M48 256c0 114.9 93.1 208 208 208s208-93.1 208-208S370.9 48 256 48 48 141.1 48 256zm32 0c0-47 18.3-91.2 51.6-124.4C164.8 98.3 209 80 256 80s91.2 18.3 124.4 51.6C413.7 164.8 432 209 432 256s-18.3 91.2-51.6 124.4C347.2 413.7 303 432 256 432s-91.2-18.3-124.4-51.6C98.3 347.2 80 303 80 256z">
      </path>
    </svg>
    <svg stroke="currentColor"
         fill="none"
         stroke-width="0"
         viewbox="0 0 15 15"
         data-on="true"
         height="1em"
         width="1em"
         xmlns="http://www.w3.org/2000/svg"
    >
      <path fill-rule="evenodd"
            clip-rule="evenodd"
            d="M3.35355 1.85355C3.54882 1.65829 3.54882 1.34171 3.35355 1.14645C3.15829 0.951184 2.84171 0.951184 2.64645 1.14645L0.646447 3.14645C0.451184 3.34171 0.451184 3.65829 0.646447 3.85355L2.64645 5.85355C2.84171 6.04882 3.15829 6.04882 3.35355 5.85355C3.54882 5.65829 3.54882 5.34171 3.35355 5.14645L2.20711 4H9.5C11.433 4 13 5.567 13 7.5C13 7.77614 13.2239 8 13.5 8C13.7761 8 14 7.77614 14 7.5C14 5.01472 11.9853 3 9.5 3H2.20711L3.35355 1.85355ZM2 7.5C2 7.22386 1.77614 7 1.5 7C1.22386 7 1 7.22386 1 7.5C1 9.98528 3.01472 12 5.5 12H12.7929L11.6464 13.1464C11.4512 13.3417 11.4512 13.6583 11.6464 13.8536C11.8417 14.0488 12.1583 14.0488 12.3536 13.8536L14.3536 11.8536C14.5488 11.6583 14.5488 11.3417 14.3536 11.1464L12.3536 9.14645C12.1583 8.95118 11.8417 8.95118 11.6464 9.14645C11.4512 9.34171 11.4512 9.65829 11.6464 9.85355L12.7929 11H5.5C3.567 11 2 9.433 2 7.5Z"
            fill="currentColor"
      >
      </path>
    </svg>
  </div>
</div>
`;

exports[`stories/components/InfiniteImageSlider LessImage smoke-test 1`] = `
<div class="_container_1r3be_1">
  <div class="_item_1r3be_11">
    <img src="/src/stories/assets/00001-2086282684.png"
         alt
    >
  </div>
  <div class="_item_1r3be_11">
    <img src="/src/stories/assets/00002-442336389.png"
         alt
    >
  </div>
  <div class="_item_1r3be_11">
    <img src="/src/stories/assets/00001-2086282684.png"
         alt
    >
  </div>
  <div class="_item_1r3be_11">
    <img src="/src/stories/assets/00002-442336389.png"
         alt
    >
  </div>
  <div class="_item_1r3be_11">
    <img src="/src/stories/assets/00001-2086282684.png"
         alt
    >
  </div>
  <div class="_item_1r3be_11">
    <img src="/src/stories/assets/00002-442336389.png"
         alt
    >
  </div>
  <div class="_item_1r3be_11">
    <img src="/src/stories/assets/00001-2086282684.png"
         alt
    >
  </div>
  <div class="_buttons_1r3be_59">
    <svg stroke="currentColor"
         fill="currentColor"
         stroke-width="0"
         viewbox="0 0 512 512"
         height="1em"
         width="1em"
         xmlns="http://www.w3.org/2000/svg"
    >
      <path d="M301.3 147.6c-7.5-7.5-19.8-7.5-27.3 0l-95.4 95.7c-7.3 7.3-7.5 19.1-.6 26.6l94 94.3c3.8 3.8 8.7 5.7 13.7 5.7 4.9 0 9.9-1.9 13.6-5.6 7.5-7.5 7.6-19.7 0-27.3l-79.8-81 81.9-81.1c7.5-7.5 7.5-19.7-.1-27.3z">
      </path>
      <path d="M256 48C141.1 48 48 141.1 48 256s93.1 208 208 208 208-93.1 208-208S370.9 48 256 48zm124.4 332.4C347.2 413.7 303 432 256 432s-91.2-18.3-124.4-51.6C98.3 347.2 80 303 80 256s18.3-91.2 51.6-124.4C164.8 98.3 209 80 256 80s91.2 18.3 124.4 51.6C413.7 164.8 432 209 432 256s-18.3 91.2-51.6 124.4z">
      </path>
    </svg>
    <svg stroke="currentColor"
         fill="currentColor"
         stroke-width="0"
         viewbox="0 0 512 512"
         height="1em"
         width="1em"
         xmlns="http://www.w3.org/2000/svg"
    >
      <path d="M210.7 147.6c7.5-7.5 19.8-7.5 27.3 0l95.4 95.7c7.3 7.3 7.5 19.1.6 26.6l-94 94.3c-3.8 3.8-8.7 5.7-13.7 5.7-4.9 0-9.9-1.9-13.6-5.6-7.5-7.5-7.6-19.7 0-27.3l79.9-81.1-81.9-81.1c-7.6-7.4-7.6-19.6 0-27.2z">
      </path>
      <path d="M48 256c0 114.9 93.1 208 208 208s208-93.1 208-208S370.9 48 256 48 48 141.1 48 256zm32 0c0-47 18.3-91.2 51.6-124.4C164.8 98.3 209 80 256 80s91.2 18.3 124.4 51.6C413.7 164.8 432 209 432 256s-18.3 91.2-51.6 124.4C347.2 413.7 303 432 256 432s-91.2-18.3-124.4-51.6C98.3 347.2 80 303 80 256z">
      </path>
    </svg>
    <svg stroke="currentColor"
         fill="none"
         stroke-width="0"
         viewbox="0 0 15 15"
         data-on="true"
         height="1em"
         width="1em"
         xmlns="http://www.w3.org/2000/svg"
    >
      <path fill-rule="evenodd"
            clip-rule="evenodd"
            d="M3.35355 1.85355C3.54882 1.65829 3.54882 1.34171 3.35355 1.14645C3.15829 0.951184 2.84171 0.951184 2.64645 1.14645L0.646447 3.14645C0.451184 3.34171 0.451184 3.65829 0.646447 3.85355L2.64645 5.85355C2.84171 6.04882 3.15829 6.04882 3.35355 5.85355C3.54882 5.65829 3.54882 5.34171 3.35355 5.14645L2.20711 4H9.5C11.433 4 13 5.567 13 7.5C13 7.77614 13.2239 8 13.5 8C13.7761 8 14 7.77614 14 7.5C14 5.01472 11.9853 3 9.5 3H2.20711L3.35355 1.85355ZM2 7.5C2 7.22386 1.77614 7 1.5 7C1.22386 7 1 7.22386 1 7.5C1 9.98528 3.01472 12 5.5 12H12.7929L11.6464 13.1464C11.4512 13.3417 11.4512 13.6583 11.6464 13.8536C11.8417 14.0488 12.1583 14.0488 12.3536 13.8536L14.3536 11.8536C14.5488 11.6583 14.5488 11.3417 14.3536 11.1464L12.3536 9.14645C12.1583 8.95118 11.8417 8.95118 11.6464 9.14645C11.4512 9.34171 11.4512 9.65829 11.6464 9.85355L12.7929 11H5.5C3.567 11 2 9.433 2 7.5Z"
            fill="currentColor"
      >
      </path>
    </svg>
  </div>
</div>
`;

exports[`stories/components/InfiniteImageSlider Normal smoke-test 1`] = `
<div class="_container_1r3be_1">
  <div class="_item_1r3be_11">
    <img src="/src/stories/assets/00001-2086282684.png"
         alt
    >
  </div>
  <div class="_item_1r3be_11">
    <img src="/src/stories/assets/00002-442336389.png"
         alt
    >
  </div>
  <div class="_item_1r3be_11">
    <img src="/src/stories/assets/00003-3958138843.png"
         alt
    >
  </div>
  <div class="_item_1r3be_11">
    <img src="/src/stories/assets/00004-15481453.png"
         alt
    >
  </div>
  <div class="_item_1r3be_11">
    <img src="/src/stories/assets/00005-15481453.png"
         alt
    >
  </div>
  <div class="_item_1r3be_11">
    <img src="/src/stories/assets/00006-15481453.png"
         alt
    >
  </div>
  <div class="_item_1r3be_11">
    <img src="/src/stories/assets/00007-15481453.png"
         alt
    >
  </div>
  <div class="_buttons_1r3be_59">
    <svg stroke="currentColor"
         fill="currentColor"
         stroke-width="0"
         viewbox="0 0 512 512"
         height="1em"
         width="1em"
         xmlns="http://www.w3.org/2000/svg"
    >
      <path d="M301.3 147.6c-7.5-7.5-19.8-7.5-27.3 0l-95.4 95.7c-7.3 7.3-7.5 19.1-.6 26.6l94 94.3c3.8 3.8 8.7 5.7 13.7 5.7 4.9 0 9.9-1.9 13.6-5.6 7.5-7.5 7.6-19.7 0-27.3l-79.8-81 81.9-81.1c7.5-7.5 7.5-19.7-.1-27.3z">
      </path>
      <path d="M256 48C141.1 48 48 141.1 48 256s93.1 208 208 208 208-93.1 208-208S370.9 48 256 48zm124.4 332.4C347.2 413.7 303 432 256 432s-91.2-18.3-124.4-51.6C98.3 347.2 80 303 80 256s18.3-91.2 51.6-124.4C164.8 98.3 209 80 256 80s91.2 18.3 124.4 51.6C413.7 164.8 432 209 432 256s-18.3 91.2-51.6 124.4z">
      </path>
    </svg>
    <svg stroke="currentColor"
         fill="currentColor"
         stroke-width="0"
         viewbox="0 0 512 512"
         height="1em"
         width="1em"
         xmlns="http://www.w3.org/2000/svg"
    >
      <path d="M210.7 147.6c7.5-7.5 19.8-7.5 27.3 0l95.4 95.7c7.3 7.3 7.5 19.1.6 26.6l-94 94.3c-3.8 3.8-8.7 5.7-13.7 5.7-4.9 0-9.9-1.9-13.6-5.6-7.5-7.5-7.6-19.7 0-27.3l79.9-81.1-81.9-81.1c-7.6-7.4-7.6-19.6 0-27.2z">
      </path>
      <path d="M48 256c0 114.9 93.1 208 208 208s208-93.1 208-208S370.9 48 256 48 48 141.1 48 256zm32 0c0-47 18.3-91.2 51.6-124.4C164.8 98.3 209 80 256 80s91.2 18.3 124.4 51.6C413.7 164.8 432 209 432 256s-18.3 91.2-51.6 124.4C347.2 413.7 303 432 256 432s-91.2-18.3-124.4-51.6C98.3 347.2 80 303 80 256z">
      </path>
    </svg>
    <svg stroke="currentColor"
         fill="none"
         stroke-width="0"
         viewbox="0 0 15 15"
         height="1em"
         width="1em"
         xmlns="http://www.w3.org/2000/svg"
    >
      <path fill-rule="evenodd"
            clip-rule="evenodd"
            d="M3.35355 1.85355C3.54882 1.65829 3.54882 1.34171 3.35355 1.14645C3.15829 0.951184 2.84171 0.951184 2.64645 1.14645L0.646447 3.14645C0.451184 3.34171 0.451184 3.65829 0.646447 3.85355L2.64645 5.85355C2.84171 6.04882 3.15829 6.04882 3.35355 5.85355C3.54882 5.65829 3.54882 5.34171 3.35355 5.14645L2.20711 4H9.5C11.433 4 13 5.567 13 7.5C13 7.77614 13.2239 8 13.5 8C13.7761 8 14 7.77614 14 7.5C14 5.01472 11.9853 3 9.5 3H2.20711L3.35355 1.85355ZM2 7.5C2 7.22386 1.77614 7 1.5 7C1.22386 7 1 7.22386 1 7.5C1 9.98528 3.01472 12 5.5 12H12.7929L11.6464 13.1464C11.4512 13.3417 11.4512 13.6583 11.6464 13.8536C11.8417 14.0488 12.1583 14.0488 12.3536 13.8536L14.3536 11.8536C14.5488 11.6583 14.5488 11.3417 14.3536 11.1464L12.3536 9.14645C12.1583 8.95118 11.8417 8.95118 11.6464 9.14645C11.4512 9.34171 11.4512 9.65829 11.6464 9.85355L12.7929 11H5.5C3.567 11 2 9.433 2 7.5Z"
            fill="currentColor"
      >
      </path>
    </svg>
  </div>
</div>
`;
